### DISCUSSION TITLE
How to pass and validate dynamic number of filter groups via query parameters?

### DISCUSSION THREAD

**Discussion by jimbotrue** [fastapi/fastapi#6480] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
There is no need for code in this example.
```

### Description

I've searched for a solution for this seemingly common problem in APIs and have not found an answer. Please point me to the right direction. I'm trying to add dynamic filtering to my api endpoint by passing groups of filters via query parameters:

Say we have a table Person:

	id
	name: str
	age: int
	height: int
	born: date

I want to create an API endpoint to get list of Persons using dynamic amount of filters

Let's say we define each filter by having three properties:

    field
    operator
    value

depending on type of field we are filtering, we'd have different operators:

	str: begins, contains, ends
	int, date: <, <=, =, >=, >

For example: 

    filters = {
        {field: "name", operator: "contains", value: "test"},
        {field: "age", operator: ">", value: 18},
        {field: "age", operator: "<=", value: 31},
        {field: "height", operator: ">", value: 150},
        {field: "height", operator: ">", value: 170},
    }

OR

    filters = {
        {field: "name", operator: "contains", value: "test"},
        {field: "age", operator: ">", value: 18},
    }

As you can see number of filters is dynamic and can include same field several times.

I need to come up with a way to convert these filters into query parameters and pass it to FastAPI

I would like to have pydantic check that "field" value of each filter group is one of those allowed

Also need to make sure that based on type of field, only appropriate operators will be used

Then I could loop through each filter and apply them to query.

How can I achieve this with FastAPI?

How would filters be represented in URI? (I would like to avoid sending them via GET request body)

Is there an alternative way to achieve this?

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.70.1

### Python Version

3.9.9

**Comment by dstlny** [comment]
> ### First Check
> * [x]  I added a very descriptive title to this issue.
> * [x]  I used the GitHub search to find a similar issue and didn't find it.
> * [x]  I searched the FastAPI documentation, with the integrated search.
> * [x]  I already searched in Google "How to X in FastAPI" and didn't find any information.
> * [x]  I already read and followed all the tutorial in the docs and didn't find an answer.
> * [x]  I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
> * [x]  I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
> * [x]  I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
> 
> ### Commit to Help
> * [x]  I commit to help with one of those options 👆
> 
> ### Example Code
> ```python
> There is no need for code in this example.
> ```
> 
> ### Description
> I've searched for a solution for this seemingly common problem in APIs and have not found an answer. Please point me to the right direction. I'm trying to add dynamic filtering to my api endpoint by passing groups of filters via query parameters:
> 
> Say we have a table Person:
> 
> ```
> id
> name: str
> age: int
> height: int
> born: date
> ```
> 
> I want to create an API endpoint to get list of Persons using dynamic amount of filters
> 
> Let's say we define each filter by having three properties:
> 
> ```
> field
> operator
> value
> ```
> 
> depending on type of field we are filtering, we'd have different operators:
> 
> ```
> str: begins, contains, ends
> int, date: <, <=, =, >=, >
> ```
> 
> For example:
> 
> ```
> filters = {
>     {field: "name", operator: "contains", value: "test"},
>     {field: "age", operator: ">", value: 18},
>     {field: "age", operator: "<=", value: 31},
>     {field: "height", operator: ">", value: 150},
>     {field: "height", operator: ">", value: 170},
> }
> ```
> 
> OR
> 
> ```
> filters = {
>     {field: "name", operator: "contains", value: "test"},
>     {field: "age", operator: ">", value: 18},
> }
> ```
> 
> As you can see number of filters is dynamic and can include same field several times.
> 
> I need to come up with a way to convert these filters into query parameters and pass it to FastAPI
> 
> I would like to have pydantic check that "field" value of each filter group is one of those allowed
> 
> Also need to make sure that based on type of field, only appropriate operators will be used
> 
> Then I could loop through each filter and apply them to query.
> 
> How can I achieve this with FastAPI?
> 
> How would filters be represented in URI? (I would like to avoid sending them via GET request body)
> 
> Is there an alternative way to achieve this?
> 
> ### Operating System
> Linux, Windows, macOS
> 
> ### Operating System Details
> _No response_
> 
> ### FastAPI Version
> 0.70.1
> 
> ### Python Version
> 3.9.9
> 
> ### Additional Context
> _No response_

Something like:

```python
from FastAPI import FastAPI
from pydantic import BaseModel, root_validator, ValidationError
import typing
import datetime

class QueryFilter(BaseModel):
       field: str
       operator: typing.Literal["includes", "contains", "<", "<=", ">=", ">"]
       value: typing.Union[str, int, datetime.date]

       @root_validator
        def validate_value(cls, values):
             
             operator, value = values.get("operator"), values.get("value")

            if not isinstance(value, (int, datetime.date)):
                 If operator in {"<=", "<", ">", ">="}: ## etc
                     raise ValueError("Some error")

              return values
              
app = FastAPI()

@app.get("/some/endpoint")
async def some_endpoint(
      *,
      filters: typing.List[QueryFilter] 
):

     ## do Something with filters
     pass
```

Written on my phone, so sorry for any syntax mistakes lol

**Comment by jimbotrue** [comment]
@dstlny your suggestion does not work, the api expects the query filters list in the Body of the request and not Query params

**Comment by dstlny** [comment]
> @dstlny your suggestion does not work, the api expects the query filters list in the Body of the request and not Query params

Changing

```python
filters: typing.List[QueryFilter] 
```

to

```python
filters: typing.List[QueryFilter] = Depends(typing.List[QueryFilter])
```

*should* do the trick iirc

**Comment by dstlny** [comment, marked as answer]
> > @dstlny your suggestion does not work, the api expects the query filters list in the Body of the request and not Query params
> 
> Changing
> 
> ```python
> filters: typing.List[QueryFilter] 
> ```
> 
> to
> 
> ```python
> filters: typing.List[QueryFilter] = Depends(typing.List[QueryFilter])
> ```
> 
> _should_ do the trick iirc

I've been thinking, and you're probably better off stringifying the list on the frontend, and then parsing it after you recieve the stringified object, rather than letting FastAPI do it. Something like"

```python
from fastapi import Depends
from pydantic import BaseModel, root_validator, parse_obj_as
import typing
import datetime
import json
from dateutil import parser

class QueryFilter(BaseModel):
	field: str
	operator: typing.Literal["includes", "contains", "<", "<=", ">=", ">"]
	value: typing.Union[str, int, datetime.date]

	@root_validator
	def validate_value(cls, values):
		
		operator, value = values.get("operator"), values.get("value")

		## is int
		if value.isdigit():
			if operator in {"<=", "<", ">", ">="}: ## etc
				pass
		else:
                       ## can't be any of these
			if operator in {"<=", "<", ">", ">="}: ## etc
				raise ValueError("Some error")

		return values

@router.get("/thing/")
async def thing(
	request: Request,
	filters: str = Query(...) ## pass as stringifed list of objects
):
	newFilter = json.loads(filters) ## load the stringified list

	all_filters = parse_obj_as(typing.List[QueryFilter], newFilter) ## let pydantic parse the JSON we just loaded... it''ll throw the same errors tha FastAPPI is expecting if it meets any errors

	print(all_filters) ## prints a list of QueryFilter

	return {}
```


Just tested this locally, stringifying from the frontend as such, and it works as it should.
FastAPI get params really aren't made for this kinda thing... :)
![image](https://user-images.githubusercontent.com/41964673/147515002-07e6b1f3-04a8-4c60-b99b-0e34e0d1025b.png)
![image](https://user-images.githubusercontent.com/41964673/147515123-a01553d1-90b6-47c6-a194-0058f74376f8.png)

**Comment by jimbotrue** [comment]
@dstlny excellent solution. Thank you very much!

**Comment by hf-krechan** [comment]
I like the idea of this setup.
But I miss the swagger documentation for the possible filters I can apply.
In the swagger, I just see filter as a query, but not the parameters/attributes I can filter.
So it is not clear for the developer what they can filter for.

![image](https://user-images.githubusercontent.com/68426071/193617986-712b91ab-afee-4da9-8a64-f0645cd1e6be.png)

Is there a way to pass the entity to the `QueryFilter` class so that its attributes are visible in the swagger documentation?

Edit:
Here the source code of my example (it is pretty much the same as above with minor fixes)

```python
from fastapi import FastAPI, Request, Query, Depends
from pydantic import BaseModel, root_validator, parse_obj_as
import typing
import datetime
import json


class QueryFilter(BaseModel):
    field: str
    operator: typing.Literal["includes", "contains", "<", "<=", ">=", ">"]
    value: typing.Union[str, int, datetime.date]

    @root_validator
    def validate_value(cls, values):

        operator, value = values.get("operator"), values.get("value")

        ## is int
        if value.isdigit():
            if operator in {"<=", "<", ">", ">="}:  ## etc
                pass
        else:
            ## can't be any of these
            if operator in {"<=", "<", ">", ">="}:  ## etc
                raise ValueError("Some error")

        return values


app = FastAPI()


@app.get("/some/endpoint")
async def thing(
    request: Request, filters: str = Query(...)  ## pass as stringifed list of objects
):
    newFilter = json.loads(filters)  ## load the stringified list

    all_filters = parse_obj_as(
        typing.List[QueryFilter], newFilter
    )  ## let pydantic parse the JSON we just loaded... it''ll throw the same errors tha FastAPPI is expecting if it meets any errors

    print(all_filters)  ## prints a list of QueryFilter

    return {}

```

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @jimbotrue  👍
